rfcs/text/????-hook-service.md

```diff
@@ -0,0 +1,134 @@
+- RFC PR: [blockprotocol/blockprotocol#430](https://github.com/blockprotocol/blockprotocol/pull/430)
+- Block Protocol Discussion: N/A
+
+# Summary
+
+[summary]: #summary
+
+This service standardises the communication between blocks and the embedding application necessary to enable embedding applications to 'take over' the rendering of parts of a block to extend functionality, while ensuring block can remain uncoupled from the embedding application, even where these are implemented by the same organisation.
+
+In order to enable direct manipulation, this service is likely to be difficult or impossible to implement in Embedding Applications which use common ‘sandboxing’ techniques including `iframe`s.
+
+# Motivation
+
+[motivation]: #motivation
+
+It is not uncommon for embedding applications to want to provide a consistent 'look and feel' across distinct blocks which implement similar features – i.e, rich text or image editing. Additionally, the implementation of these features may require integrating with systems bespoke to the embedding application. We don't want blocks to have to 'know' about the implementation of this functionality, or necessarily the shape of the data involved in providing this functionality. Therefore, the embedding application needs to be responsible for providing this functionality, but this depends on co-operation by block authors. 
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A 'hook' is an injection point blocks provide to embedding applications to allow them to optionally render or modify already rendered views associated with a path to data provided by the graph service's `blockEntity` property. 
+
+The 'optionally' here is key. Blocks cannot require embedding applications to implement a handler for any specific 'hook'. Therefore, each 'hook' must specify a fallback rendering strategy which embedding applications implementing the hook service can use instead of implementing the hook themselves.   
+
+A classical example would be rich text editing. In this case, you may not want to provide the underlying data storing the formatted text, as this may be bespoke to your embedding application. But you may still want to provide the plain text value for the block to make use of – i.e, if the block provides a word count feature.
+
+A block implements a hook by sending a `hook` message to the embedding application, with five pieces of information:
+- `node: HTMLElement | null`: The DOM node the embedding application will render a view into, or modifying an already rendered view. 
+- `type: "text" | "image" | "video" | string`: The type of view for the hook. While this can be any string as a fallback must be passed if `type` is not implemented, we believe `text`, `image`, and `video` will be commonly used. Straying from these commonly used types will make your block less portable between different embedding applications, particularly those designed to support all blocks.  
```

Comment:
how about `number`?

do we need to distinguish between `string` and something like `rich-text`?

Use case for these additions: to allow EAs to use their own inputs for common form inputs.

Reply:
I think we need to be clear what we're talking about when we say "type".

We're not (in my mind) talking about data types, like strings, or rich-text -- but the type of _interface_ a user interacts with while dealing with data in these nodes.

They probably will have different interfaces for inputting plain strings vs WYSIWYG-style rich text.

Reply:
This was discussed internally at https://hashintel.slack.com/archives/C02LG39FJAU/p1658246784302989 as something which is out of scope for this RFC.

Reply:
To be clear/for public record, the thing we've decided will be out-of-scope here is "constraints on hooks" (e.g. how do I selectively disable specific functions within a text-editing interface that may or may not be provided by an embedding application, within a particular hook). One option for how this will work is that function names will be arrived at in a non-prescriptive manner, and supported as independently/organically as the hook `type` names themselves. We can establish precedents in our original implementations without binding future users.